### PR TITLE
 🏃[KCP] add semver check to kubeproxy upgrade

### PR DESCRIPTION
**Pull request by sedefsavas**
**What this PR does / why we need it**:
Add semver version check to  kube-proxy upgrade

/assign @vincepri

### CODE REVIEW

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 1]
```diff
@@ -589,17 +590,49 @@ func (w *Workload) UpdateKubeProxyImageInfo(ctx context.Context, kcp *controlpla
 	if len(ds.Spec.Template.Spec.Containers) == 0 {
 		return nil
 	}
-	newImageName, err := util.ModifyImageTag(ds.Spec.Template.Spec.Containers[0].Image, kcp.Spec.Version)
+
+	parsedImage, err := reference.ParseNormalizedNamed(ds.Spec.Template.Spec.Containers[0].Image)
```
Comment (detiber) [inline controlplane/kubeadm/internal/workload_cluster.go]: Potential issue for followup: this should likely iterate over the containers like the coreDNS implementation is doing.

**Comment by ncdc** [comment]
This ideally should be in the KCP webhook validation for `spec.version`

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 2]
```diff
@@ -589,17 +590,49 @@ func (w *Workload) UpdateKubeProxyImageInfo(ctx context.Context, kcp *controlpla
...
+	}
+
+	if fromParsedVersion == toParsedVersion {
+		return nil
+	}
+
+	newImageName := fmt.Sprintf("%s:%s", imageRepository, toParsedVersion)
+
+	patch := client.MergeFrom(ds.DeepCopy())
+	ds.Spec.Template.Spec.Containers[0].Image = newImageName
```
Comment (detiber) [inline controlplane/kubeadm/internal/workload_cluster.go]: Potential issue for followup: This should likely iterate over the containers similar to the coreDNS implementation.

**Comment by ncdc** [comment]
(the semver check, that is)

**Comment by ncdc** [comment]
I'm not sure we need all the logic in here if we validate the kcp.spec.version is a valid semver in the webhook, but maybe I'm missing something?

**Comment by sedefsavas** [comment]
Adding semver check during validation will allow both v1.16.6 and 1.16.6 so we need another modification inside controller.

**Comment by sedefsavas** [comment]
Changed this to follow how UpdateKubernetesVersionInKubeadmConfigMap() and ReconcileKubeletRBACRole() is consuming version. 

Sending semver version to UpdateKubeProxyImageInfo().

**Comment by vincepri** [comment]
@sedefsavas we also parse the version in `upgradeControlPlane`, I'm wondering if we can actually make it parse before calling both of these and reuse it in both places? We can pass it in as an extra argument for now

**Comment by sedefsavas** [comment]
@ vincepri
closing this PR in regards to https://github.com/kubernetes-sigs/cluster-api/pull/2691
It fixes "v" issue.

**Comment by vincepri** [comment]
Sounds good, let's close this one for now

/close

**Pull request closed without merging**